Refactor FXIOS-13400 [Homepage Layout] Cache section measurements to avoid redundant recalculations

Aniket-pd opened the pull request:
## :scroll: Tickets  
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-13400)  
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/29124)

## :bulb: Description  
This PR introduces caching for homepage layout measurements to reduce redundant recalculations:  

- **Added `LayoutMeasurementCache` helper** to store previously computed section measurements (Top Sites, Stories, Search Bar).  
- **Normalized width-dependent cache keys** so identical inputs reuse cached values instead of rebuilding views.  
- **Introduced keyed header-height caching** and shared tallest-cell results for Stories so both section layout and spacer sizing reuse the same computation.  
 

**Motivation:** Homepage refreshes often repeat the same state (same width, same visibility). Without caching, expensive layout work and view instantiation happens repeatedly. With this change, redundant work is skipped, improving CPU efficiency and reducing layout thrash.


<details>
this PR only affects internal performance 
</details>

## :pencil: Checklist  
- [x] I filled in the ticket numbers and a description of my work  
- [x] I updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)  
- [ ] I ensured unit tests pass and wrote tests for new code  
- [ ] If working on UI, I checked and implemented accessibility (Dynamic Text and VoiceOver)  
- [ ] If adding telemetry, I read the [data stewardship requirements](https://github.com/mozilla-mobile/firefox-ios/wiki/Adding-Glean-Telemetry-Events) and will request a data review  
- [ ] If adding or modifying strings, I read the [guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/How-to-add-and-modify-Strings) and will request a string review from l10n  
- [ ] If needed, I updated documentation and added comments to complex code

Aniket-pd commented:
Hey, I accidentally closed this earlier ,my bad!
I’ve reopened it now. @ih-codes @Foxbolts could you please give it a review when you have a moment?

Foxbolts reviewed (commented):
Hey @Aniket-pd, thank you so much for tackling this issue. I think you have the right approach with caching to avoid redundant calculations! 

An issue I ran into with your changes is referencing cached tallest story cell heights on dynamic type changes where the cell heights should theoretically be changing and recalculated (see screenshots below)

This also results in section heights being incorrect, causing the homepage to scroll unnecessarily (something we are trying to avoid by having the `spacer` section)

Other than that, this looks great, though additionally, I would add a few comments describing your cache system.

Thanks again, and let me know if you need any further clarifications to make those changes!

| Actual | Expected |
| ------------- | ------------- |
| <img width="1206" height="2622" alt="Simulator Screenshot - iPhone 17 - 2025-09-29 at 16 02 34" src="https://github.com/user-attachments/assets/29b5c5b8-5792-4fba-b5b2-2c556ca09de7" /> | <img width="1206" height="2622" alt="Simulator Screenshot - iPhone 17 - 2025-09-29 at 16 04 06" src="https://github.com/user-attachments/assets/2e387cc7-6940-490b-81da-68d85f913121" /> |

Foxbolts reviewed (approved):
This looks and works well for me, lets run CI - tagging @cyndichin for another pair of 👀

cyndichin commented on firefox-ios/Client/Frontend/Home/Homepage Rebuild/Layout/HomepageSectionLayoutProvider.swift:
```diff
@@ -128,6 +191,8 @@ final class HomepageSectionLayoutProvider: FeatureFlaggable {
     private var logger: Logger
     private var windowUUID: WindowUUID
+    private var measurementsCache = LayoutMeasurementCache()
+    private var headerHeightCache: [HeaderMeasurementKey: CGFloat] = [:]
```
We can move this `headerHeightCache` into under the `HomepageMeasumerCache` and in its own file instead, we can retrieve using `measurementsCache.getHeaderHeight`

cyndichin commented on firefox-ios/Client/Frontend/Home/Homepage Rebuild/Layout/HomepageSectionLayoutProvider.swift:
```diff
@@ -630,14 +771,73 @@ final class HomepageSectionLayoutProvider: FeatureFlaggable {
+        return result
+    }
+
+    private func normalizedDimension(_ value: CGFloat) -> Double {
```
can we add a comment here?

cyndichin reviewed (commented):
thanks @Aniket-pd for tackling this ticket!! I left some comments to abstract out some of this logic from being in the Section Layout class and keep it isolated. I would love to see tests around this area and it would be easier if we can test on the logic in isolation and only expose necessary methods here. Let me know if you need more clarity on this! 

I haven't pulled down to test since @Foxbolts vetted and can let QA review. As long as everything seems to work fine in these following cases:
- iPad (split view)
- iPhone
- Landscape / Portrait
- Dynamic Text

cyndichin commented on firefox-ios/Client/Frontend/Home/Homepage Rebuild/Layout/HomepageSectionLayoutProvider.swift:
```diff
@@ -253,7 +318,11 @@ final class HomepageSectionLayoutProvider: FeatureFlaggable {
             collectionViewWidth: environment.container.contentSize.width
         )
-        let tallestCellHeight = getTallestStoryCellHeight(cellWidth: cellWidth)
+        let storiesMeasurement = getStoriesMeasurement(
+                    environment: environment,
+                    cellWidth: cellWidth
+                )
+        let tallestCellHeight = storiesMeasurement.tallestCellHeight
```
It seems the only usage of the `getStoriesMeasurement` is to retrieve the tallest height, so maybe we can encapsulate the logic in its own file similar to `HomepageDimensionCalculator` and then we can pass in whats needed like `environment, cell width, storiesState` to keep this file small and potentially add code coverage for the other file.

Aniket-pd commented:
Thanks for the suggestion @cyndichin! 
I’ve refactored the cache logic as discussed ,moved LayoutMeasurementCache into its own file (HomepageLayoutMeasurementCache.swift) and encapsulated the set/get logic inside methods.
Also updated all references in the layout provider to use the new structure.
